website/views.py

```diff
@@ -670,36 +725,59 @@ def question_delete(request, question_id):
                 question.body,
                 delete_reason,
             )
-            email = EmailMultiAlternatives(
-                subject, '',
-                sender_email, to,
-                bcc=[bcc_email],
-                headers = {"Content-type":"text/html;charset=iso-8859-1"}
-            )
-            email.attach_alternative(message, "text/html")
-            email.content_subtype = 'html'  # Main content is text/html
-            email.mixed_subtype = 'related'
-            email.send(fail_silently = True)
-            #send_mail(subject, message, sender_email, to, fail_silently = True)
+            mail_uids = to_uids(question)
+            for uid in mail_uids:
+                to = [get_user_email(uid)]
+                send_email(sender_email, to, subject, message, bcc_email)
+
+    question.is_active = False
+    question.save()
+    answers = question.answer_set.all()
+    for answer in answers:
+        answer.is_active = False
+        answer.save()
+        for comment in answer.answercomment_set.all():
+            comment.is_active = False
+            comment.save()
+    return render(request,
+                  'website/templates/question-delete.html',
+                  {'title': title})
+
 
-    question.delete()
-    return render(request, 'website/templates/question-delete.html', {'title': title})
+@login_required
+def question_restore(request, question_id):
+    question = get_object_or_404(Question, id=question_id, is_active=False)
+    if not is_moderator(request.user, question) or not settings.MODERATOR_ACTIVATED:
+        return render(request, 'website/templates/not-authorized.html')
+    question.is_active = True
+    question.save()
+    for answer in question.answer_set.all():
+        answer.is_active = True
+        answer.save()
+        for comment in answer.answercomment_set.all():
+            comment.is_active = True
+            comment.save()
+    return HttpResponseRedirect('/question/{0}/'.format(question_id))
 
 # View for deleting answer, notification is sent to person who posted answer
+# @user_passes_test(is_moderator)
 @login_required
-@user_passes_test(is_moderator)
 def answer_delete(request, answer_id):
 
-    answer = get_object_or_404(Answer, id = answer_id)
+    answer = get_object_or_404(Answer, id=answer_id)
     question_id = answer.question.id
 
-    if (request.method == "POST"):
+    if ((request.user.id != answer.uid or AnswerComment.objects.filter(answer=answer,
+                                                                       is_active=True).exclude(uid=answer.uid).exists()) and (not is_moderator(request.user, answer.question) or not settings.MODERATOR_ACTIVATED)):
+        return render(request, 'website/templates/not-authorized.html')
+
+    if (request.method == "POST") and (settings.MODERATOR_ACTIVATED):
```

Comment:
is this MODERATOR_ACTIVATED check required again? all authorization checks are already done.

Reply:
Have you checked this one?

Reply:
this is required because mail is sent only when moderator deletes the answer.
So When authorised user deletes the answer, mail is not sent to anyone.
User can only be authorized to delete an answer if he has posted this answer and no comment is being posted under this answer.